### PR TITLE
escape-html: set up autogenerated API docs

### PR DESCRIPTION
**Pull request by oandregal**
**Related to** https://github.com/WordPress/gutenberg/issues/14227

This PR sets up auto-generated API docs for the `escape-html` package through the `npm run docs:build` command and updates its README file.

## Test instructions

- Delete or modify the contents at `packages/escape-html/README.md` within the start `<!-- START TOKEN(Autogenerated API docs) -->` and end `<!-- END TOKEN(Autogenerated API docs) -->` tokens.
- Execute `npm run docs:build` and verify they are re-created.
- Review that the API docs match the JSDoc of the exported symbols.

### CODE REVIEW

**Review by gziolo — APPROVED** [review]
So nice so nice. For the record, I'm only checking output generated. If it would fail at any point we will have to fix it globally 😅

**Comment by nosolosw** [comment]
Have to rebase from `master` after merging every PR due to conflicts in the package array :/ Will merge when CI tests pass.

**Comment by aduth** [comment]
> Have to rebase from `master` after merging every PR due to conflicts in the package array :/ Will merge when CI tests pass.

Could we find some way to automate this process? This seems like yet another array which needs to be updated when creating a new package. For context, historically it's been a big pain to create new packages just because you needed to update several arrays of every package name throughout the project. We've done (and should continue to do) well to limit this by leveraging discovery as much as possible.

Could it be enough to [glob](https://www.npmjs.com/package/glob) `'./packages/*/README.md` and include any package which contains a "start token" ? Or just run it on all packages (what's the downside? Is it something we plan to enable in the future anyways?)

**Comment by nosolosw** [comment]
> Could it be enough to glob './packages/*/README.md and include any package which contains a "start token" ? Or just run it on all packages (what's the downside? Is it something we plan to enable in the future anyways?)

I've been reluctant to implement this for two reasons:

- I feared that many packages will require different configuration. Of the 27 PRs that I've prepared only `rich-text` needed a different set of `--ignore` rules. So I was proved wrong in this one. We will still need to keep a list for these, though.

- A second concern that I have is this: what should be the result of `docgen <input-file> --output <output-file> --to-token` if no token is found in the file? I think it should fail (exit code different from 0 + error message shown). The auto-discover mechanism for Gutenberg could interfere with that API; as a result, we may miss that something is wrong if someone would delete the token tags by accident, for example.

I'm happy to revisit this decission if it becomes a big concern.

A different point is that the current mechanism requires a bit of a whac-a-mole game _for me_ when it comes to merge the PRs, though. I'm not a fan of this either! 1547e5d fixes it. Being a obvious change that only adds comments I'm going to abuse Greg trust and will go ahead and merge this :)

**Comment by aduth** [comment]
> * I feared that many packages will require different configuration. Of the 27 PRs that I've prepared only `rich-text` needed a different set of `--ignore` rules. So I was proved wrong in this one. We will still need to keep a list for these, though.

I don't think the package-specific configuration should occur in this script, though. I could imagine something like a configuration file within the package, if it were to come to that. Something like [`.npmrc`](https://docs.npmjs.com/files/npmrc), discovered by the script.

>A second concern that I have is this: what should be the result of docgen <input-file> --output <output-file> --to-token if no token is found in the file? I think it should fail (exit code different from 0 + error message shown). The auto-discover mechanism for Gutenberg could interfere with that API; as a result, we may miss that something is wrong if someone would delete the token tags by accident, for example.

I don't really see an issue just having it be effectively a noop if there's no token found. How does it work today? Are you suggesting that a secondary purpose of the script is to validate the input? I can see how there might be some value in that, I suppose. But I could still see maybe some additional option to just allow a silent failure if we're expecting some inputs to maybe not contain the token.

None of this applies though if we just search the file contents for the presence of the token before running them through docgen.

>I'm happy to revisit this decission if it becomes a big concern.

In isolation, none of the other "arrays of all packages" seemed like a big issue, but the cumulative of them is what led to the degraded developer experience. I'd rather we be proactive about avoiding a return to that.

Example from a much earlier pull request, where each of these has been since (mostly) automated away:

- https://github.com/WordPress/gutenberg/pull/7018/files#diff-6b0584709544530e1ccbee1bf679be9aR9
- https://github.com/WordPress/gutenberg/pull/7018/files#diff-11e9f7f953edc64ba14b0cc350ae7b9dR159
- https://github.com/WordPress/gutenberg/pull/7018/files#diff-e4403a877d80de653400d88d85e4801aR41
- https://github.com/WordPress/gutenberg/pull/7018/files#diff-620130744b6b73b822a3df609671d930R119

**Comment by gziolo** [comment]
> I don't think the package-specific configuration should occur in this script, though. I could imagine something like a configuration file within the package, if it were to come to that. Something like .npmrc, discovered by the script.

It might make sense to put some additional info in either `package.json` or something like `.wprc` file where we could provide some additional rules specific to a given package. I like this idea, at the moment we make some decision based on the root `package.json` file when deciding which packages build for Gutenberg plugin:

https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L26-L28

However, we have to maintain some lists of exceptions in a few places:

https://github.com/WordPress/gutenberg/blob/master/webpack.config.js#L77-L83
https://github.com/WordPress/gutenberg/blob/master/test/unit/jest.config.json#L4

I guess those are also some places where `docgen` might need some special handling.

File: bin/update-readmes.js
Comment (aduth) [inline bin/update-readmes.js]: Couple thoughts:

- If this is a regular expression, should we be assuring `^` so that we'd not inadvertently ignore the token `definitely_not_unstable` (ignoring the fact this would, for other reasons, not be allowed)
- The [guidelines](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-apis) clearly prescribe the double-underscore `__` prefix. I don't think we want the tolerance here. I want to see clear and loudly in the produced documentation if we're not naming these correctly.

Comment (nosolosw) [reply bin/update-readmes.js]: Yeah, I wanted a bit of leeway on purpose. There are a couple of offenders right now on `master`. Created https://github.com/WordPress/gutenberg/issues/14337 to review this after all packages are set up.

**Pull request merged**
